### PR TITLE
New Http Rest Client

### PR DESCRIPTION
**Pull request by sjpadgett**
- implements OAuth
- implements guzzle

### CODE REVIEW

**Comment by sjpadgett** [comment]
Porting Publish now to use this which will give some clarity for how to use. Generally oeHttp is static meaning creating an instance is not necessary. Requests and responses are instantiated by oeHttp for you. So oeHttp::get(uri) or oeHttp::usingHeaders($headers)->post(url) is generally all that is needed for a request.
For Fhir & OAuth could be something like:
oeHttp::usingAuthEndpoint($endpoint)->withCredentials(user,pass,default)->get($uri . 'Patient/221')
and on.
More to come but thought i'd get this up for comments and question.

File: composer.json [hunk 1]
```diff
@@ -36,7 +36,9 @@
         "vlucas/phpdotenv" : "2.5.1",
         "yubico/u2flib-server" : "1.0.2",
         "zendframework/zendframework" : "2.5.3",
-        "particle/validator": "2.3.3"
+        "particle/validator": "2.3.3",
+        "guzzlehttp/guzzle": "6.3.3",
+        "kamermans/guzzle-oauth2-subscriber": "^1.0.5"
```
Comment (bradymiller) [inline composer.json]: better to just strictly hard-code the version:
```
"kamermans/guzzle-oauth2-subscriber": "1.0.5"
```
(ie. drop the carat)

Comment (sjpadgett) [reply composer.json]: Thought I had thus the 1.0.5. Chances are i'm sleep developing again:) will change.

**Comment by sjpadgett** [comment]
Also, you know me, always the rebel. you say support PHP >= 7 so, there you are:) I'd like to say it was just to remind you to remove check from Travis but, at last, i'm not that smart....

**Comment by sjpadgett** [comment]
Redo of OAuth init. 
        $uri = "Patient/30";
        $endpoint = "http://localhost/openemr/apis/fhir/auth";
        $response = oeHttp::withOAuth($credentials,$endpoint)->get($uri);
        $returned = json_decode($response->body()) ''; // patient resource
then following calls against new token:
$response = oeHttp::reAuth()->usingApiBaseUri("http://localhost/openemr/apis/fhir")->get('Encounter');
Or:
$response = oeHttp::reAuth()->get('http://localhost/openemr/apis/fhir/Encounter?patient=30');
And so on.

**Comment by sjpadgett** [comment]
@bradymiller This client works out very well. Makes using rest api's very simple. Check how was used in ClientAppController.

File: API_README.md [hunk 2]
```diff
@@ -13,10 +13,15 @@ This project aims to provide an easy-to-use JSON-based REST API for OpenEMR's mo
 - [@bradymiller](https://github.com/bradymiller)
 
 ### Prerequsite
-Enable this API service in OpenEMR menu: Administration->Globals->Connectors->"Enable OpenEMR REST API"
+Enable this API service in OpenEMR menu: Administration->Globals Connectors tab.
```
Comment (bradymiller) [inline API_README.md]: Do you mind if I revert the change above? :)

File: API_README.md [hunk 3]
```diff
@@ -13,10 +13,15 @@ This project aims to provide an easy-to-use JSON-based REST API for OpenEMR's mo
...
+
+### Using API Internally
+There are several ways to make API calls from an authorized session and maintain security:
+* cURL or Guzzle requests
+* oeHttp OpenEMR Http Rest Client
 
 ### Endpoints
-Note: FHIR endpoints follow normal FHIR REST endpoints. Use `https://domain/apis/fhir as base URL.`
+Note: FHIR endpoints follow normal FHIR REST endpoints. Use `https://domain/apis/fhir as base UR.`
```
Comment (bradymiller) [inline API_README.md]: missing L

File: _rest_config.php [hunk 4]
```diff
@@ -153,6 +153,24 @@ function destroySession()
...
+                return $post_json;
+            } else {
+                parse_str($post_data, $post_variables);
+                if (count($post_variables)) {
+                    return $post_variables;
+                }
+            }
+        }
+            // Not able to retrieve POST data, return FALSE.
+            return false;
```
Comment (bradymiller) [inline _rest_config.php]: indentation off in above 2 lines (yes, that is likely only going to be my only input on this review :) )

**Comment by bradymiller** [comment]
Code looks good to me (i only noted a couple very, very, very minor things)
Now I am going to try this out

**Comment by bradymiller** [comment]
Wow, this is really cool. Just did a bunch of curl commandline calls for api and fhir on the demo here: https://www.open-emr.org/wiki/index.php/Development_Demo#With_demo_data

Note I haven't gotten my local instance working (to test out this code), but will sort that out later. I say bring this in :) Nice work!

**Comment by sjpadgett** [comment]
Hopefully the framework gods can make peace with this:) I've heard from Tiago of MiNA and he has been able to get a container version of OpenEMR up and running and was able to authenticate and retrieve a Fhir Patients bundle via their client. I will need to implement Authorization grant type and get on with writing more resources but, I think we finally have the piece in place.
Onward to Glory. (trying out a new battle cry)...
